### ISSUE TITLE
Model Builder feature wish list

### ISSUE THREAD

**Issue by kaihuchen** [tensorflow/tfjs-core#33]
I am studying the possibility of building some practical client-side ML solutions with deeplearnjs, and the [Model Builder](https://pair-code.github.io/deeplearnjs/demos/model-builder/model-builder-demo.html) is an excellent place for me to start. However, there are still some critical features needed for my purpose that I do not see on the roadmap. So here is a list of my wishlist items for the team.

- Support a feature for saving and restoring ANY trained model. 
- Adding support for a test phase, i.e., the ability to load a specific dataset and test it against a trained model. If there is no near-term plan to support this, then I'd appreciate some pointers as to where I should look into in order for me to hack it out by myself.
- During the test phase, support adding a user-specified callback function to be triggered by test result (e.g., if any class receives higher than 90% confidence then call the function). Again, not sure how useful this is to others, so pointers on how to achieve this in deeplearnjs is much appreciated.

**Comment by kaihuchen** [comment]
One more item:

- Support more types of layers in the **deferred execution model**, so that it is possible to support popular models such as the Squeezenet. The Imagenet demo does show a squeezenet using the **immediate execution model**, but it is currently impossible to duplicate it in the Model Builder for lacking a 'fire' layer type.

**Comment by nsthorat** [comment]
Okay, so I have some good news and some bad news.

I think we will in the near future spend some time making the model builder a little bit more extensible. It's was our first demo, and honestly wasn't written the way I would write it if I knew it would be a larger scale product.

> Support a feature for saving and restoring ANY trained model.
> Adding support for a test phase, i.e., the ability to load a specific dataset and test it against a trained model. If there is no near-term plan to support this, then I'd appreciate some pointers as to where I should look into in order for me to hack it out by myself.

So luckily, we had built weights saving as a shim to TensorFlow, but we decided to turn it off for the launch. However, the code is still there :) This is a little long-winded, but if you're up for it, here it goes:

Check out this function:
https://github.com/PAIR-code/deeplearnjs/blob/master/demos/model-builder/model-builder.ts#L815

This should support getting weights *into* the browser (the button that this gets triggered from is hidden, you will have to unhide it). However, we don't have the other direction, getting weights *out* of the browser. In theory, this should be pretty straightforward, if you want to take it on (I will be happy to review your code).

You will have to add a method that dumps a LayerWeightsDict and triggers a download. The model builder owns a list of model-layers, each of which own a "LayerBuilder". LayerBuilder can return the weights for its own operation - you will have to pass the Session object down into the LayerBuilder, through the ModelLayer.

Once you have the session in LayerBuilder, you can look up the value of the variable  from "Session.activationArrayMap", which is a class that look up an NDArray value from a Tensor id (you need the Tensor ids of variables). Some of the variables are constructed directly in the LayerBuilder, so you just need to save the id of those tensors (property id of the Tensor object returned by g.Variable)

But in the fully connected layer, we use a g.layers abstraction which creates variables internally. They have names that are deterministic, so you have to find the Tensor id programatically.

The graph object has a method, "getNodes", which will return an array of Node objects. These represent the nodes in the differentiable graph. Each Node object has a "name" -- you can see what they look like here:
https://github.com/PAIR-code/deeplearnjs/blob/master/src/graph_layers.ts#L32

Now that you have the Node object for the Variable in the dense layer, you can get the Tensor id output by "node.tensor.id". Phew, now you can look it up on the Session object to get the weights.

I know this is super clunky. We're going to make this easier soon, and probably make model builder a little better, however, contributions are totally welcome, and I can help you with the process.

> During the test phase, support adding a user-specified callback function to be triggered by test result (e.g., if any class receives higher than 90% confidence then call the function). Again, not sure how useful this is to others, so pointers on how to achieve this in deeplearnjs is much appreciated.

Here is the method that gets called when we render the logits (these are the softmax values that you can check):
https://github.com/PAIR-code/deeplearnjs/blob/master/demos/model-builder/model-builder.ts#L685

Keep in mind, we are sampling a subset of the test data when rendering. If you want to infer through the entire dataset and get a callback, this requires a little bit of work, and might be really slow.

> Support more types of layers in the deferred execution model, so that it is possible to support popular models such as the Squeezenet. The Imagenet demo does show a squeezenet using the immediate execution model, but it is currently impossible to duplicate it in the Model Builder for lacking a 'fire' layer type.

Unfortunately, SqueezeNet requires a real Graph. The model builder only lets you build a sequential models, that is, all data flows in and out of layers. SqueezeNet has branching convolutions that we wouldn't be able to express in the model builder, unfortunately ('fire module' is a set of convolutions). For graph based models, you should really just try to use the underlying methods the model builder users without using the heavy duty UI. All of the tools the model builder uses should be straightforward to use (xhr-dataset with normalization, graphs, session, etc).

PHEW.

Sorry I'm a little late responding. I'm currently vacationing in Australia :)

**Comment by kaihuchen** [comment]
@nsthorat   Much appreciated for the detailed information!  Please do not rush your replies while you are on your vacation, or else I am going to feel guilty!

Regarding my request about saving weights, the first half of your explanation looks easy enough. The second half seems kind of complex since it goes pretty deep into the bowel of deeplearnjs. I may give it a stab time permitting.          

Regarding my question about the 'callback during test phase', from inspecting model-builder.ts I now see that it automatically allocates one sixth of the given dataset for testing. What I had in mind is actually about a separate test phase **after** the training is all done. You can think of it as a kind of  **product phase** where a trained model is put to some practical use (say, watching a garden through a webcam to detect birds), and it is at that point that I am looking for a callback mechanism, so there is performance issue there. Your reply looks good, but I guess for that I will need to get the said post-training test phase working first.

**Comment by nsthorat** [comment]
Ahh I see what you are saying, you want a way to deploy a trained model that was used from the model builder.

We're going to soon implement a way to serialize a model entirely for deployment, and we'll make sure the model builder is hooked up to it. I'm back from vacation on Thursday, and we have a new person starting on Monday project so we'll probably prioritize this. Will get back to you soon!

**Comment by reiinakano** [comment]
Hello @nsthorat, I wasn't sure if I should start a new issue about this, so I'll ask my question here. Above, you stated 

> I think we will in the near future spend some time making the model builder a little bit more extensible. It's was our first demo, and honestly wasn't written the way I would write it if I knew it would be a larger scale product.

May I know what things you'd redo if you were to write it again?

I'm currently planning to write something quite similar to model-builder, and was wondering if I should just use your code directly as a template and change things around to fit my use case. I'm not experienced with Polymer and am now debating whether to just rewrite from scratch with React (which I have a bit of experience with) or learn Polymer and reuse some of the components you've built here already. I've gone deep into my research and even wound up at this thread, https://news.ycombinator.com/item?id=13159251, where surprisingly, yours was the top post (which made me a bit more confused since you don't seem to like Polymer). I'm leaning towards sticking with Polymer and reappropriating the relevant code for model-builder, so I want to know which parts you'd keep/throw away.

Ultimately, though, I'm not experienced enough with JS in general to make an informed decision here. Would appreciate any feedback. Thanks!

**Comment by nsthorat** [comment]
That's excellent!

I've never used React but I've heard good things. The biggest thing I would do is *not* use Polymer. I definitely advise against it, it's honestly just a nightmare to develop with.

One thing I want to let you know is that we're going to change our high level differentiation API soon. No more Graphs / Sessions, we're going to be more like the pytorch define by run (we'll only have the NDArrayMath layer & auto differentiation). That being said, you can start building this out, but just make sure you have this in mind. The data API is probably going to change at some point soon as well (XHRDataset, etc).

**Comment by reiinakano** [comment]
Wow, that is a BIG change. I'm glad to hear it, though, I love Pytorch and honestly can't stand Tensorflow. Any idea when that'll be rolling out?

Thanks for the answer! Guess I'll do a quick PoC first while waiting for the changes.

**Comment by nsthorat** [comment]
Hopefully in the coming couple months. It's a huge change, and we're going to have to do it gradually. It's only 2 of us at Google full time right now so we're trying to get help :)

The graph / session model are horribly unintuitive, I agree. We only decided to go that route to mirror TensorFlow, but now we're regretting it since people have a hard time grasping it. The graph / session are also impossible to debug, even for those of us that wrote the original code. The new world will actually make stack traces sane.

But yeah go ahead and build a proof of concept, but I'd just keep this in mind!

**Comment by frqc** [comment]
Hi I'm also very interested in the BIG CHANGE of applying "define by run" like pytorch. Do you have any roadmap or anything others may contribute to?

**Comment by nsthorat** [comment]
We're going to write a design doc soon. There's a core piece of the puzzle we need to get done first, which is decoupling the NDArray class from its storage mechanism, which will make the define-by-run differentiation much simpler, and more importantly backend agnostic. We'll make sure to keep github updated once we write that doc.

**Comment by andrii-maglovanyi** [comment]
Could you give some insight on the status of model export/import functionality please?

**Comment by erikwilson** [comment]
I took the past couple days to create serialization and deserialization routines for graphs. It is probably hacky, doesn't have proper tests, and is written in ES6 for node.js, so would need refactoring to end up in main branch, but here you go: https://gist.github.com/erikwilson/770668bdb82eeed7b7f0ec33b9dbe051

**Comment by andrii-maglovanyi** [comment]
@erikwilson great job!

**Comment by erikwilson** [comment]
Thanks! I re-published the code on npm with some better tests:
https://www.npmjs.com/package/deeplearn-graph-serializer

**Comment by QuantumInformation** [comment]
I wouldn't worry about too much the react vs polymer vs X, the debate never ends.

**Comment by nsthorat** [comment]
We're going to stop work on the current model builder. We will be building another one that uses the layers API and is built a little better in the coming months, so stay tuned.
